src/main/java/net/revelc/code/apilyzer/maven/plugin/AnalyzeMojo.java

```diff
@@ -264,32 +339,145 @@ public URL apply(String input) {
             }
           }
         });
-    cl = new URLClassLoader(urls.toArray(new URL[0]));
+    cl = new URLClassLoader(urls.toArray(new URL[0]), null);
     return ClassPath.from(cl);
   }
 
+  private Annotation[] getAnnotations(ClassInfo classInfo) {
+    if (classInfo.getName().startsWith("com.sun") || classInfo.getName().startsWith("java.")) {
+      // was getting class not found exceptions when trying to get annotations for com.sun class...
+      return new Annotation[0];
+    }
+
+    return getAnnotations(classInfo.load());
+  }
+
+  private Annotation[] getAnnotations(Class<?> clazz) {
+    return clazz.getDeclaredAnnotations();
+  }
+
+  /**
+   * Builds the set of Types that are in the public API.
+   */
   private void buildPublicSet(ClassPath classPath, List<Class<?>> publicApiClasses,
       TreeSet<String> publicSet) {
-    for (ClassInfo classInfo : classPath.getAllClasses()) {
+    classLoop: for (ClassInfo classInfo : classPath.getAllClasses()) {
       // TODO handle empty includes case; maybe?
-      for (String includePattern : includes) {
-        if (classInfo.getName().matches(includePattern)) {
-          boolean exclude = false;
-          for (String excludePattern : excludes) {
-            if (classInfo.getName().matches(excludePattern)) {
-              exclude = true;
-              break;
+
+      // Do this check before possibly attempting any annotation checks as these require class
+      // loading. If the class is excluded by a pattern, then no need to load class.
+      if (patternExcludes(classInfo)) {
+        continue;
+      }
+
+      Annotation[] annotations;
+      if (includeAnnotations.size() > 0 || excludeAnnotations.size() > 0) {
+        annotations = getAnnotations(classInfo);
+      } else {
+        annotations = new Annotation[0];
+      }
+
+
+      for (String pattern : includeAnnotations) {
+        for (Annotation annotation : annotations) {
+          if (annotation.toString().matches(pattern)) {
```

Comment:
This would probably be faster if the pattern was compiled just once outside the inner for loop.

Reply:
Yeah I was thinking that to, but didn't have time.  We should open an issue.